### ISSUE TITLE
Excess properties in tagged union not checked if type is optional and literal

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#31655]
**TypeScript Version:**  3.4.1, 3.5.1

**Search Terms:** tagged union optional literal excess properties props

**Code**

```ts
type Button = { tag: 'button'; type?: 'submit'; };
type Anchor = { tag: 'a'; type?: string; href: string };

type Union = Button | Anchor;
const obj: Union = {
    tag: 'button',
    type: 'submit',

    // no error, bad! :-(
    href: 'foo',
    // error, good! :-)
    href2: 'foo',
};
```

If I replace `Button` with:

``` ts
type Button = { tag: 'button'; type?: string };
```

That is, if I replace the optional string literal with an optional string, then it works as expected (I get an error for the excess prop `href`), suggesting this is down to the fact that we're using string literals?

**Comment by OliverJAsh** [comment]
Even using `href: never` on `Button` doesn't help:

``` ts
type Button = { tag: 'button'; type?: 'button'; href: never; };
```

**Comment by RyanCavanaugh** [comment]
This works as expected if you turn strict null checks off 🙃

**Comment by sandersn** [comment]
Right now the discriminable property check gives up if more than one type in the union matches *any* discriminant property. Unfortunately, `type: 'submit'` matches both Button.type and Anchor.type.

I'm going to try making the check stricter and see if it breaks anything.

Not sure why strictNullChecks makes this work.
